### PR TITLE
Android | Visited items: Create edit mode

### PR DESCRIPTION
**Pull request by gclssvglx**
# Android | Visited items: Create edit mode

This change adds the edit screen for visited items. 

It also fixes a padding issue [identified here](https://govukverify.atlassian.net/browse/GOVUKAPP-645) on the view screen.

## Note 🎵 

Visited items on the edit screen retain  the link colouration for the title and `open in...` icon from the view screen. However, the link is no longer clickable and the alt text for the icon has been removed. This is in-line with the iOS implementation.

This change does not address:

- Removing visited items from the Realm database
- Selecting and deselecting all visited items
- Removing all visited items
- The `Select all` and `Remove` button logic 

All of the above will be addressed in [this ticket](https://govukverify.atlassian.net/browse/GOVUKAPP-652).

## JIRA ticket(s)
  - [GOVAPP-651](https://govukverify.atlassian.net/browse/GOVUKAPP-651)

## Figma
  - [Figma board](https://www.figma.com/design/5u5cvVCgKY3Rng2ADNV30K/2024-07---GOV.UK-app---beta-designs?node-id=1-83650&node-type=canvas&t=2TPX3uBvnxrgiSto-0)

## Screenshots

| Light | Dark |
| --- | --- |
| after-light-view | after-dark-view |
| after-light-edit | after-dark-edit |

### CODE REVIEW

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/VisitedScreen.kt [hunk 1]
```diff
@@ -57,22 +71,64 @@ private fun VisitedScreen(
...
+                    .padding(bottom = GovUkTheme.spacing.small)
+            )
+        }
+        LazyColumn(
+            Modifier.padding(horizontal = GovUkTheme.spacing.medium)
+                .padding(top = GovUkTheme.spacing.small)
+        ) {
             item {
-                val visitedItems = uiState?.visited
                 if (visitedItems != null) {
```
Comment (TimGJD) [inline feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/VisitedScreen.kt:132]: I know this unchanged but we could do like I've mentioned elsewhere whilst we're here and change to?:

`visitedItems ?: return@item`

Or use `visitedItems?.let { .. }`

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        },
+        bottomBar = {
+            BottomNavBar(
+                selectText = selectText,
+                removeText = removeText,
+                modifier = modifier
+            )
+        }
+    ) { innerPadding ->
+        LazyColumn (
```
Comment (TimGJD) [inline feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt:113]: Noticed a couple of places need formatting in this file so suggest a 'Option + Command + L' to correct

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+                Checkbox(
+                    checked = checked,
+                    onCheckedChange = {
+                        checked = it
+
+                        if (checked) {
+                            println("Adding : $title [$url]")
+                        } else {
+                            println("Removing : $title [$url]")
+                        }
```
[the remark this answers is not in the document]
Comment (gclssvglx) [reply feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt:185]: Yeah, just validating that this works really. Happy to remove, but will be removed in the edit storage ticket anyway

[the remark this answers is not in the document]
Comment (gclssvglx) [reply feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt]: Removed

File: design/src/main/kotlin/uk/govuk/app/design/ui/theme/Color.kt [hunk 4]
```diff
@@ -15,6 +15,7 @@ private val Blue6 = Color(0xFF7AC2FF)
 private val Blue7 = Color(0xFFE8F1F8)
 
 private val Yellow = Color(0xFFFFDD00)
+private val Red1 = Color(0xFFD4351C)
```
[the remark this answers is not in the document]
Comment (gclssvglx) [reply design/src/main/kotlin/uk/govuk/app/design/ui/theme/Color.kt:19]: Fixed up

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+                .padding(horizontal = GovUkTheme.spacing.medium)
+                .fillMaxWidth()
+                .background(GovUkTheme.colourScheme.surfaces.background)
+        ) {
+            item {
+                if (visitedItems != null) {
+                    val lastVisitedText = stringResource(R.string.visited_items_last_visited)
+
+                    visitedItems.forEach { (sectionTitle, visitedItems) ->
+                        if (visitedItems.isNotEmpty()) {
```
[the remark this answers is not in the document]
Comment (gclssvglx) [reply feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt]: Went with `let` as suggested [here](https://github.com/alphagov/govuk-mobile-android-app/pull/117#discussion_r1843793043)

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    ) { innerPadding ->
+        LazyColumn (
+            modifier = modifier
+                .padding(innerPadding)
+                .padding(horizontal = GovUkTheme.spacing.medium)
+                .fillMaxWidth()
+                .background(GovUkTheme.colourScheme.surfaces.background)
+        ) {
+            item {
+                if (visitedItems != null) {
```
[the remark this answers is not in the document]
Comment (gclssvglx) [reply feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt]: Went with `let`

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/VisitedScreen.kt [hunk 7]
```diff
@@ -57,22 +71,64 @@ private fun VisitedScreen(
...
     val title = stringResource(R.string.visited_items_title)
+    val editText = stringResource(R.string.visited_items_edit_button)
 
     Column(modifier) {
-        ChildPageHeader(
-            text = title,
-            onBack = onBack
-        )
-        LazyColumn(Modifier.padding(horizontal = GovUkTheme.spacing.medium)) {
+        val visitedItems = uiState?.visited
```
[the remark this answers is not in the document]
Comment (gclssvglx) [reply feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/VisitedScreen.kt]: Fixed up

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/VisitedScreen.kt (hunk 1 above)
Comment (gclssvglx) [reply feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/VisitedScreen.kt]: Went with `isNullOrEmpty()` - seemed nicer and easier to read

File: feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt (hunk 2 above)
Comment (gclssvglx) [reply feature/visited/src/main/kotlin/uk/govuk/app/visited/ui/EditVisitedScreen.kt]: Fixed up

**Review by TimGJD — APPROVED** [review]

**Pull request merged**